pull-request.sh

```diff
@@ -58,7 +60,9 @@ create_pull_request() {
 
     # Check if the branch already has a pull request open
     DATA="{\"base\":${TARGET}, \"head\":${SOURCE}, \"body\":${BODY}}"
-    RESPONSE=$(curl -sSL -H "${AUTH_HEADER}" -H "${HEADER}" --user "${GITHUB_ACTOR}" -X GET --data "${DATA}" ${PULLS_URL})
+    if ! RESPONSE=$(safe_curl -X GET --data "${DATA}" ${PULLS_URL}); then
```

Comment:
Could you make this condition more explicit instead of just using the ! sign? It's not obvious what safe_curl returns, and we should (via the code) show what we are explicitly looking for to determine a failing condition.

Reply:
The way that I have written this is the only idiom I'm aware of for handling this situation.

* The `$(..)` opens a new subshell with the set -e behavior of the current shell.
* safe_curl then inherits that, so the `-f` flag to curl causes the `$(..)` subshell to exit with an error code if curl exits
* The `if` statement in bash _only_ checks the return code of expressions it is passed. So in terms of "it's not clear what safe_curl returns", it is always the return code of the functions inside of it. So this checks the result of doing `curl` because of the error propagation -- this is the same way that `test`/`[` work. See my second option below for a way we could make this more explicit.
* If a command fails in the top-level script (for example. if you have `RESPONSE=$(cmd)` on a line by itself then the script will exit because of `set -e`

It *could* be rewritten as:

```bash
set +e
RESPONSE=$(safe_curl ...)
RETVAL=$?
set -e
```

The reason that retval was always being set to `0` before was because `-f` wasn't getting passed to curl, which meant that http errors never caused it to fail, meaning that basically never failed.

Or I could change `safe_curl` to:

```bash
do_curl() {
   set +e
   curl ...
   return $?
}
```

To make it more explicit?

Reply:
Definitely! I think we should take the second approach, because other than being more explicit, we can then still echo the result to the user, and set the variables in the workflow step, before acting on them.

Reply:
This second version is exactly the same semantics as the current version, is that what you're referring to?

Reply:
And I believe that every code path right now does explicitly print the return code so that users can always see it?

Reply:
The snippets I'm referring to are of this type:
```
echo ::set-env name=ASSIGNEES_RETURN_CODE::${RETVAL}
echo ::set-env name=ASSIGNEES_RETURN_CODE::${RETVAL}
echo ::set-output name=assignees_return_code::${RETVAL}
echo ::set-output name=assignees_return_code::${RETVAL}
```
We want to make sure these variables are set (and printf's printed to the user) before we take action. It looks like you fixed this so I don't think this is an issue anymore.